lib/construct.ts

```diff
@@ -428,15 +428,16 @@ export class Node {
    * Invokes "prepare" on all constructs (depth-first, post-order) in the tree under `node`.
    */
   public prepare() {
-    const constructs = this.findAll(ConstructOrder.PREORDER);
+    let constructs = this.findAll(ConstructOrder.PREORDER);
 
     // Aspects are applied root to leaf
     for (const construct of constructs) {
       Node.of(construct).invokeAspects();
     }
 
-    // Use .reverse() to achieve post-order traversal
-    for (const construct of constructs.reverse()) {
+    // Since constrcuts can be added to the tree when invoking an Aspect, recreate the constructs list
+    constructs = this.findAll(ConstructOrder.POSTORDER);
+    for (const construct of constructs) {
```

Comment:
Can we keep the `PREORDER` and `.reverse` call here? So, this would be:

```ts
for (const construct of this.findAll(ConstructOrder.PREORDER).reverse()) {
  // ...
}
```

My justification: keep the behavior as close as to what it was before (even if it's not strictly correct).

Reply:
Yeah we can, reverse(preorder) is also a topological sorting, just like postorder, so both are **right**. 
My motivation for changing it was to make the code clearer and easier to debug, easier in the sense that postorder of a tree is easier to think of than the reverse of preorder. 
That being said, this change does breaks several tests in the CDK, most of them are easy to fix as they shouldn't have taken a dependency on the order of the construct tree.
@eladb @rix0rrr what do you think?

Reply:
It depends which tests are broken. Can you submit a PR with those changes so we can evaluate?

Reply:
Lets keep the current order, I'll open an issue in the CDK repo

Reply:
@eladb see https://github.com/aws/aws-cdk/issues/7059

Reply:
preserved the previous order and opened https://github.com/aws/aws-cdk/issues/7059